rclpy/rclpy/qos_event.py

```diff
@@ -215,18 +225,36 @@ def create_event_handlers(
                 callback=self.liveliness,
                 event_type=QoSSubscriptionEventType.RCL_SUBSCRIPTION_LIVELINESS_CHANGED,
                 parent_handle=subscription_handle))
+
         if self.incompatible_qos:
             event_handlers.append(QoSEventHandler(
                 callback_group=callback_group,
                 callback=self.incompatible_qos,
                 event_type=QoSSubscriptionEventType.RCL_SUBSCRIPTION_REQUESTED_INCOMPATIBLE_QOS,
                 parent_handle=subscription_handle))
+        else:
+            # Register default callback when not specified
+            try:
+                event_type = QoSSubscriptionEventType.RCL_SUBSCRIPTION_REQUESTED_INCOMPATIBLE_QOS
+                event_handlers.append(QoSEventHandler(
+                    callback_group=callback_group,
+                    callback=self._default_incompatible_qos_callback,
+                    event_type=event_type,
+                    parent_handle=subscription_handle))
+            except UnsupportedEventTypeError:
+                self._logger.warn(
+                    'This rmw implementation does not support ON_REQUESTED_INCOMPATIBLE_QOS '
+                    'events, you will not be notified when Subscriptions request an incompatible '
+                    'QoS profile from Publishers on the same topic.',
+                    once=True)
         return event_handlers
 
 
 class PublisherEventCallbacks:
     """Container to provide middleware event callbacks for a Publisher."""
 
+    _logger = rclpy.logging.get_logger('SubscriptionEventCallbacks')
```

Comment:
Is this the convention, where the logger is specific to the class instead of the class instance?

Reply:
It was the best way I could think of to warn once per process run, rather than for every publisher/subscription. I believe `once=True` applies to the logger instance, so I created one for the class rather than the instance.

Reply:
Oh but the name is definitely not correct. will fix

Reply:
Should the logger used in the default callback function be the same as the one used during the registering of the default callback function be the same? We should get more guidance from OSRF.